File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.cpp
class Trie {
private:
    struct Node{
        Node *chld[26]{} ;
        bool is_word;
        Node():is_word(false){
            fill(chld,chld+26, nullptr) ;
        }
    };
    Node *root = new Node() ;
public:
    void insert(string word) {
        Node *cr = root ;
        for(auto c : word){
            if(cr->chld[c - 'a'] == nullptr){
                cr->chld[c - 'a'] = new Node() ;
            }
            cr = cr->chld[c - 'a'] ;
        }
        cr->is_word = true ;
    }
    
    bool search(string word) {
        Node *cr = root ;
        for(auto c : word){
            if(cr->chld[c - 'a'] == nullptr) return false ;
            cr = cr->chld[c - 'a'] ;
        }
        return cr->is_word ;
    }
    
    bool startsWith(string prefix) {
        Node *cr = root ;
        for(auto c : prefix){
            if(cr->chld[c - 'a'] == nullptr) return false ;
            cr = cr->chld[c - 'a'] ;
        }
        
        return true ;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.java
class Trie {
    Node root ;
    private class Node{
        Node[] chld ;
        int cnt;
        Node(){
            chld = new Node[26];
            cnt = 0 ;
        }
    }
    public Trie() {
        root = new Node();
    }
    
    public void insert(String word) {
        Node current = root ;
        for(char c : word.toCharArray()){
            if(current.chld[c - 'a'] == null){
                current.chld[c - 'a'] = new Node();
            }
            current = current.chld[c - 'a'];
        }
        current.cnt ++ ;
    }
    
    public boolean search(String word) {
        Node current = root ;
        for(char c : word.toCharArray()){
            if(current.chld[c - 'a'] == null){
                return false ;
            }
            current = current.chld[c - 'a'];
        }
        return (current.cnt > 0);
    }
    
    public boolean startsWith(String prefix) {
        Node current = root ;
        for(char c : prefix.toCharArray()){
            if(current.chld[c - 'a'] == null){
                return false ;
            }
            current = current.chld[c - 'a'];
        }
        return true ;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */